File: bot/supabase.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any
from bot.config import SupabaseSettings
from bot.http import JsonHttpClient
from bot.models import ContentItem, Interaction


class SupabaseClient:
    def __init__(self, settings: SupabaseSettings, http: JsonHttpClient | None = None) -> None:
        self.base_url = settings.url.rstrip("/")
        self.http = http or JsonHttpClient()
        self.headers = {
            "apikey": settings.service_role_key,
            "Content-Profile": "public",
            "Accept-Profile": "public",
        }
        if not settings.service_role_key.startswith("sb_secret_"):
            self.headers["Authorization"] = f"Bearer {settings.service_role_key}"

    def _rpc(self, function_name: str, payload: Mapping[str, Any]) -> Any:
        return self.http.request(
            "POST",
            f"{self.base_url}/rest/v1/rpc/{function_name}",
            headers=self.headers,
            json_body=dict(payload),
        )

    def claim_interactions(
        self,
        *,
        batch_size: int = 10,
        max_attempts: int = 5,
        stale_lock_minutes: int = 10,
    ) -> list[Interaction]:
        data = self._rpc(
            "claim_interactions",
            {
                "batch_size": batch_size,
                "max_attempts": max_attempts,
                "stale_lock_minutes": stale_lock_minutes,
            },
        )
        if not isinstance(data, list):
            raise RuntimeError("claim_interactions returned an unexpected response")
        return [Interaction.from_record(item) for item in data if isinstance(item, dict)]

    def claim_due_content(
        self,
        *,
        batch_size: int = 5,
        max_attempts: int = 5,
        stale_lock_minutes: int = 15,
    ) -> list[ContentItem]:
        data = self._rpc(
            "claim_due_content",
            {
                "batch_size": batch_size,
                "max_attempts": max_attempts,
                "stale_lock_minutes": stale_lock_minutes,
            },
        )
        if not isinstance(data, list):
            raise RuntimeError("claim_due_content returned an unexpected response")
        return [ContentItem.from_record(item) for item in data if isinstance(item, dict)]

    def _update(self, table: str, row_id: str, values: Mapping[str, Any]) -> None:
        self.http.request(
            "PATCH",
            f"{self.base_url}/rest/v1/{table}",
            headers={**self.headers, "Prefer": "return=minimal"},
            query={"id": f"eq.{row_id}"},
            json_body=dict(values),
        )

    def update_interaction(self, interaction_id: str, values: Mapping[str, Any]) -> None:
        self._update("interactions", interaction_id, values)

    def update_content(self, content_id: str, values: Mapping[str, Any]) -> None:
        self._update("content_queue", content_id, values)

    def mark_interaction_failed(self, interaction_id: str, error: str, max_attempts: int = 5) -> None:
        self._rpc(
            "mark_interaction_failed",
            {"p_id": interaction_id, "p_error": error[:4_000], "p_max_attempts": max_attempts},
        )

    def mark_content_failed(self, content_id: str, error: str, max_attempts: int = 5) -> None:
        self._rpc(
            "mark_content_failed",
            {"p_id": content_id, "p_error": error[:4_000], "p_max_attempts": max_attempts},
        )

    def insert_keyword_interaction(self, record: Mapping[str, Any]) -> None:
        self.http.request(
            "POST",
            f"{self.base_url}/rest/v1/interactions",
            headers={
                **self.headers,
                "Prefer": "resolution=ignore-duplicates,return=minimal",
            },
            query={"on_conflict": "source_item_id"},
            json_body=dict(record),
        )

File: supabase/functions/_shared/supabase.ts
import type { ContentRow, InteractionRow } from "./types.ts";

interface RequestOptions {
  method?: string;
  body?: unknown;
  prefer?: string;
}

function errorMessage(body: string): string {
  try {
    const parsed = JSON.parse(body) as { code?: string; message?: string };
    return [parsed.code, parsed.message].filter(Boolean).join(": ").slice(0, 1000);
  } catch {
    return body.slice(0, 1000) || "Unknown Supabase error";
  }
}

export class SupabaseRestClient {
  constructor(
    private readonly baseUrl: string,
    private readonly apiKey: string,
  ) {}

  private async request<T>(path: string, options: RequestOptions = {}): Promise<T> {
    const headers = new Headers({
      apikey: this.apiKey,
      "content-type": "application/json",
      "accept-profile": "public",
      "content-profile": "public",
    });
    if (!this.apiKey.startsWith("sb_secret_")) {
      headers.set("authorization", `Bearer ${this.apiKey}`);
    }
    if (options.prefer) headers.set("prefer", options.prefer);

    let response: Response;
    try {
      response = await fetch(`${this.baseUrl.replace(/\/$/, "")}/rest/v1/${path}`, {
        method: options.method ?? "GET",
        headers,
        body: options.body === undefined ? undefined : JSON.stringify(options.body),
        signal: AbortSignal.timeout(30_000),
      });
    } catch {
      throw new Error("Supabase request failed");
    }

    const raw = await response.text();
    if (!response.ok) {
      throw new Error(`Supabase ${response.status}: ${errorMessage(raw)}`);
    }
    return (raw ? JSON.parse(raw) : undefined) as T;
  }

  private rpc<T>(name: string, body: Record<string, unknown>): Promise<T> {
    return this.request<T>(`rpc/${name}`, { method: "POST", body });
  }

  claimInteractions(batchSize: number, maxAttempts: number): Promise<InteractionRow[]> {
    return this.rpc<InteractionRow[]>("claim_interactions", {
      batch_size: batchSize,
      max_attempts: maxAttempts,
      stale_lock_minutes: 10,
    });
  }

  claimDueContent(batchSize: number, maxAttempts: number): Promise<ContentRow[]> {
    return this.rpc<ContentRow[]>("claim_due_content", {
      batch_size: batchSize,
      max_attempts: maxAttempts,
      stale_lock_minutes: 15,
    });
  }

  updateInteraction(id: string, values: Record<string, unknown>): Promise<void> {
    return this.request<void>(`interactions?id=eq.${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: values,
      prefer: "return=minimal",
    });
  }

  updateContent(id: string, values: Record<string, unknown>): Promise<void> {
    return this.request<void>(`content_queue?id=eq.${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: values,
      prefer: "return=minimal",
    });
  }

  markInteractionFailed(id: string, error: string, maxAttempts: number): Promise<void> {
    return this.rpc<void>("mark_interaction_failed", {
      p_id: id,
      p_error: error.slice(0, 4000),
      p_max_attempts: maxAttempts,
    });
  }

  markContentFailed(id: string, error: string, maxAttempts: number): Promise<void> {
    return this.rpc<void>("mark_content_failed", {
      p_id: id,
      p_error: error.slice(0, 4000),
      p_max_attempts: maxAttempts,
    });
  }

  insertKeywordInteraction(values: Record<string, unknown>): Promise<void> {
    return this.request<void>("interactions?on_conflict=source_item_id", {
      method: "POST",
      body: values,
      prefer: "resolution=ignore-duplicates,return=minimal",
    });
  }
}
